text/0000-partial_types.md

```diff
@@ -0,0 +1,672 @@
+- Feature Name: `partial_types`
+- Start Date: 2023-04-18
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+
+# Summary
+[summary]: #summary
+
+Partial types proposal is a generalization on "partial borrowing"-like proposals (more correct name is "partial not borrowing" or "qualified borrowing" since Rust allows partial borrowing already).
+
+This proposal is a universal roadmap "how to do partial not consumption (including partial not borrowing) right", and not under the hood of the Rust compiler.
+
+Partial Types is a **minimal** and full extension to the Rust Type System, which allows to safe control easily partial parameters and all kinds of partial not consumption.
+
+Advantages: maximum type safety, maximum type control guarantee, no ambiguities, flexibility, usability and universality.
+
+
+# Motivation
+[motivation]: #motivation
+
+Safe, Flexible controllable partial parameters for functions and partial not consumption (including partial not borrowing) are highly needed and this feature unlock huge amount of possibilities.
+
+Partial borrowing is already possible in Rust, as partial referencing and partial moves.
+
+But partial parameters are forbidden now, as qualified consumption: partial not borrowing, partial not referencing, partial not moving and partial initializing.
```

Comment:
I feel like the motivation for partial borrows is very clear, but the motivation for this implementation of it is not. "Unlocking huge amounts of possibilities" doesn't seem terribly convincing to me; for example, we *could* implement the duck-typing templates system from C++ to unlock loads of possibilities, but we don't because it's clunky, errors at monomorphisation time by design, and encourages bad code. What real cases does this feature unlock that make it feel worthwhile?

Reply:
First, this proposal almost remain style of using Rust as already is (by contrast to duck-typing).
1) It is full backward-compatible.  This proposal DO NOT change or block any existed feature! Before and after implementation.
2) It adds some **safe** flexibility to **safe** code by **safe** methods.
3) Simplicity in binary - Type integrity just say by type to compiler, that some fields are forbidden to use _for everyone ever_. And that allows to use ordinary references as "partial" and ordinal variables as "partial". No extra actions with variables or pointers are needed.
4) Any partial type error is a compiler error, all types are erased after type-check, so no extra-cost in binary is needed.
5) it is universal rule - that mean minimal special cases on implementation.
6) It is minimal universal-extension - all other proposals propose less than this with  more or same cost